vm/qemu/qmp.go

```diff
@@ -0,0 +1,108 @@
+package qemu
+
+import (
+	"fmt"
+	"net"
+	"encoding/json"
+	"errors"
+)
+type qmpVersion struct {
+	Package string
+	QEMU    struct {
+		Major int
+		Micro int
+		Minor int
+	}
+}
+
+type qmpBanner struct {
+        QMP struct {
+		Version qmpVersion
+        }
+}
+
+type qmpCommand struct {
+	Execute   string      `json:"execute"`
+	Arguments interface{} `json:"arguments,omitempty"`
+}
+
+type hmpCommand struct {
+	Command string `json:"command-line"`
+	CPU     int    `json:"cpu-index"`
+}
+
+type qmpResponse struct {
+	Error string       `json:"error,omitempty"`
+	Return interface{} `json:"return,omitempty"`
+}
+
+func (inst *instance) qmp_check_connection() error {
+	if inst.mon != nil {
+		return nil
+	}
+
+	addr := fmt.Sprintf("127.0.0.1:%v", inst.monport)
+	conn, err := net.Dial("tcp", addr)
```

Comment:
We seem to be leaking this socket, we need to close it somewhere.

Reply:
if `conn` is referenced by `instance`, does not it get freed by the GC when `instance` is freed?

Reply:
Yes, it is. net.Conn should even have finalizer that will close the fd.
However, non-memory resources don't provide GC pressure, so we can e.g. run out of ports/fds before GC runs.
In Go if a type has Close, it's supposed to be called synchronously. Also provides a way to check for errors (Close errors are frequently ignore, but sometimes they are not).

Reply:
inst.Close will be a handy point to close the conn.